app/master/build_request_handler.py

```diff
@@ -73,6 +75,22 @@ def _build_preparation_loop(self):
         """
         while True:
             build = self._request_queue.get()
+            project_id = build.project_type.project_id()
+
+            if project_id not in self._project_preparation_locks:
+                self._project_preparation_locks[project_id] = Lock()
+
+            project_lock = self._project_preparation_locks[project_id]
+            async_preparation_thread = SafeThread(target=self._prepare_build_async, name='BuildPreparationThread',
+                                                  args=[build, project_lock])
+            async_preparation_thread.start()
+
+    def _prepare_build_async(self, build, project_lock):
+        """
+        :type build: Build
+        :type project_lock: Lock
+        """
```

Comment:
Can you explain what is happening in the below line?

Is the lock acquired and released via the `with` keyword?  What happens if multiple threads call `acquire()`?  Do they acquire it in the order that they called it, or do we even care?